File: app/ios/NeuroPulse/Models/OTAModels.swift
import Foundation

// OTA firmware update state model.
// Dual-bank OTA: hub has Bank A (current) and Bank B (staging).
// Per NP-FW-EMMC-001 Rev A §8 and BLE OTA protocol spec.

// MARK: - OTA phase (matches hub OTA_STATUS notify byte 0)

enum OTAPhase: UInt8, CustomStringConvertible {
    case idle           = 0x00
    case preparing      = 0x01  // Hub clearing Scratch partition
    case transferring   = 0x02  // Receiving firmware chunks
    case verifying      = 0x03  // Hub computing SHA-256 + Ed25519 check
    case verified       = 0x04  // Signature OK; awaiting COMMIT opcode from app
    case applying       = 0x05  // Writing inactive bank + staging boot swap + about to reset
    case complete       = 0x06  // Post-reboot, running from new bank
    case failed         = 0xFF

    init(rawPacketByte: UInt8) {
        self = OTAPhase(rawValue: rawPacketByte) ?? .idle
    }

    var description: String {
        switch self {
        case .idle:         return "Ready"
        case .preparing:    return "Preparing update…"
        case .transferring: return "Transferring firmware…"
        case .verifying:    return "Verifying signature…"
        case .verified:     return "Signature verified"
        case .applying:     return "Applying update — do not disconnect"
        case .complete:     return "Update complete"
        case .failed:       return "Update failed"
        }
    }

    var isTerminal: Bool { self == .complete || self == .failed }
    var isBusy: Bool {
        switch self {
        case .idle, .complete, .failed: return false
        default: return true
        }
    }
}

// MARK: - Firmware version (semantic versioning, Comparable)

struct FirmwareVersion: Comparable, Equatable, Hashable, CustomStringConvertible {
    let major: Int
    let minor: Int
    let patch: Int

    init?(_ string: String) {
        let parts = string.split(separator: ".").compactMap { Int($0) }
        guard parts.count == 3,
              parts[0] >= 0, parts[1] >= 0, parts[2] >= 0 else { return nil }
        (major, minor, patch) = (parts[0], parts[1], parts[2])
    }

    // Parse from hub GATT firmware_version characteristic (uint32, little-endian).
    // Encoding: bits [23:16] = major, [15:8] = minor, [7:0] = patch.
    init?(gattBytes: Data) {
        guard gattBytes.count >= 4 else { return nil }
        let raw = gattBytes.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: 0, as: UInt32.self) }
        let littleEndian = UInt32(littleEndian: raw)
        major = Int((littleEndian >> 16) & 0xFF)
        minor = Int((littleEndian >> 8)  & 0xFF)
        patch = Int( littleEndian        & 0xFF)
    }

    var description: String { "\(major).\(minor).\(patch)" }

    static func < (lhs: Self, rhs: Self) -> Bool {
        if lhs.major != rhs.major { return lhs.major < rhs.major }
        if lhs.minor != rhs.minor { return lhs.minor < rhs.minor }
        return lhs.patch < rhs.patch
    }
}

// MARK: - Firmware manifest (decoded from distribution endpoint)

struct FirmwareManifest: Decodable {
    let version: String
    let releaseDate: String         // ISO 8601 date string, e.g. "2026-06-09"
    let imageSizeBytes: Int
    let sha256Hex: String           // Expected SHA-256 of the firmware binary
    let ed25519KeyFingerprint: String
    let isSafetyMCUFirmware: Bool
    let releaseNotes: String

    func toFirmwareImage() -> FirmwareImage? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        guard let date = formatter.date(from: releaseDate) else { return nil }
        return FirmwareImage(
            version: version,
            buildDate: date,
            imageSizeBytes: imageSizeBytes,
            ed25519PublicKeyFingerprint: ed25519KeyFingerprint,
            isSafetyMCUFirmware: isSafetyMCUFirmware,
            releaseNotes: releaseNotes
        )
    }
}

// MARK: - Firmware image (display model)

struct FirmwareImage: Identifiable {
    var id: UUID = UUID()
    var version: String
    var buildDate: Date
    var imageSizeBytes: Int
    var ed25519PublicKeyFingerprint: String  // hex — shown to user before confirmation
    var isSafetyMCUFirmware: Bool           // requires explicit separate user confirmation
    var releaseNotes: String
}

// MARK: - OTA session (transfer progress tracking)

struct OTASession {
    var image: FirmwareImage
    var startedAt: Date = Date()
    var completedAt: Date?

    // Chunk-level tracking
    var totalChunks: Int = 0
    var sentChunks: Int = 0

    // Byte-level tracking (ISC-110: bytes/total bytes progress bar)
    var totalBytes: Int = 0
    var sentBytes: Int = 0

    // ATT_MTU(512) − 3B ATT header − 1B OTA opcode − 2B chunk index = 506 max.
    // Using 496 leaves a safe margin for GATT stack overhead variations.
    static let chunkSize = 496

    var isActive: Bool { completedAt == nil && sentBytes > 0 }
    var bytesProgress: Double { totalBytes > 0 ? Double(sentBytes) / Double(totalBytes) : 0 }

    static func formattedBytes(_ bytes: Int) -> String {
        // Integer round-half-up avoids banker's rounding producing "0 KB" for sub-1 KB values.
        let kb = (bytes + 512) / 1024
        if kb < 1000 { return "\(max(1, kb)) KB" }
        return String(format: "%.1f MB", Double(bytes) / (1024.0 * 1024.0))
    }
}

File: app/android/core/src/main/kotlin/com/neurone/core/models/OtaModels.kt
package com.neurone.core.models

// Port of app/ios/NeurOne/Models/OTAModels.swift.
// Wire values are frozen — hub firmware contract (CLAUDE.md §13.5 OTA entry).

enum class OtaPhase(val rawValue: Int) {
    IDLE(0x00),
    RECEIVING(0x01),
    VERIFYING(0x02),
    COMMITTING(0x03),
    VERIFIED(0x04),
    REBOOTING(0x05),
    COMPLETE(0x06),
    FAILED(0xFF);

    val isBusy: Boolean
        get() = this == RECEIVING || this == VERIFYING || this == COMMITTING || this == REBOOTING

    val isTerminal: Boolean
        get() = this == COMPLETE || this == FAILED

    companion object {
        fun from(rawValue: Int): OtaPhase? = entries.firstOrNull { it.rawValue == rawValue }
    }
}

data class OtaStatusPacket(
    val phaseRaw: Int,
    val progressPercent: Int,
    val errorCode: Int,
) {
    val isError: Boolean get() = errorCode != 0
    val phase: OtaPhase? get() = OtaPhase.from(phaseRaw)
}

// FIRMWARE_VERSION GATT encoding: uint32 little-endian —
// bits [23:16]=major, [15:8]=minor, [7:0]=patch.
data class FirmwareVersion(
    val major: Int,
    val minor: Int,
    val patch: Int,
) : Comparable<FirmwareVersion> {

    override fun compareTo(other: FirmwareVersion): Int =
        compareValuesBy(this, other, { it.major }, { it.minor }, { it.patch })

    override fun toString(): String = "$major.$minor.$patch"

    companion object {
        /** Parse "1.2.3". Returns null on any malformed input. */
        fun parse(versionString: String): FirmwareVersion? {
            val parts = versionString.split(".")
            if (parts.size != 3) return null
            val nums = parts.map { it.toIntOrNull() ?: return null }
            if (nums.any { it < 0 }) return null
            return FirmwareVersion(nums[0], nums[1], nums[2])
        }

        /** Parse the 4-byte little-endian GATT payload. */
        fun fromGattBytes(data: ByteArray): FirmwareVersion? {
            if (data.size < 4) return null
            val raw = (data[0].toLong() and 0xFF) or
                ((data[1].toLong() and 0xFF) shl 8) or
                ((data[2].toLong() and 0xFF) shl 16) or
                ((data[3].toLong() and 0xFF) shl 24)
            return FirmwareVersion(
                major = ((raw shr 16) and 0xFF).toInt(),
                minor = ((raw shr 8) and 0xFF).toInt(),
                patch = (raw and 0xFF).toInt(),
            )
        }
    }
}

// In-flight OTA transfer bookkeeping. Chunk payload size matches the iOS app
// and hub firmware: 496 bytes of firmware data per BLE write (2B index + data).
data class OtaSession(
    val totalBytes: Long,
    val sentBytes: Long = 0,
) {
    val progressFraction: Double
        get() = if (totalBytes <= 0) 0.0 else sentBytes.toDouble() / totalBytes.toDouble()

    companion object {
        const val CHUNK_SIZE: Int = 496

        /**
         * Human-readable byte count with integer round-half-up KB/MB —
         * mirrors iOS `formattedBytes`, which avoids banker's-rounding
         * "0 KB" for sub-kilobyte transfers.
         */
        fun formattedBytes(bytes: Long): String {
            if (bytes < 1024) return "$bytes B"
            val kb = (bytes + 512) / 1024      // round half up
            if (kb < 1024) return "$kb KB"
            val mb = (kb + 512) / 1024
            return "$mb MB"
        }
    }
}
